[JIT] Refactor TupleType to take a FunctionSchema

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21839 [JIT] NamedTuple serialization
* **#21836 [JIT] Refactor TupleType to take a FunctionSchema**
* #21835 [JIT] Introduce NamedType
* #21813 [JIT] Some small fixes for NamedTuple

Differential Revision: [D15860003](https://our.internmc.facebook.com/intern/diff/D15860003)

Commit:
[JIT] Refactor TupleType to take a NamedTupleSpec

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -846,11 +847,34 @@ struct CAFFE2_API FutureType : public SingleElementType<TypeKind::FutureType, Fu
+      createFunctionSchema();
+    }
+
+    std::vector<TypePtr> types;
```
This class needs to be encapsulated. Public members will not age well. Also it seems to have redundant info for instance a FunctionSchema and a list of types and names. That is exactly what an Argument list stores. I'd prefer we just kept a FunctionSchema.

zdevito commented on torch/csrc/jit/python_ir.cpp:
```diff
@@ -678,6 +678,12 @@ void initPythonIRBindings(PyObject* module_) {
         }
         return types;
       });
+  py::class_<TupleType::NamedTupleSpec>(m, "_NamedTupleSpec")
```
This appears unused. Let's not bind something into python if it isn't used there. It is hard to prevent users from using it if it is there.

zdevito reviewed (commented):
Refactor looks good but NamedTupleSpec needs better encapsulation.

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a NamedTupleSpec"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a FunctionSchema"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a FunctionSchema"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a FunctionSchema"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head

Commit:
Update on "[JIT] Refactor TupleType to take a FunctionSchema"

[JIT] Refactor TupleType to take a NamedTupleSpec

gh-metadata: pytorch pytorch 21836 gh/jamesr66a/4/head
